pkg/parser/parser.go

```diff
@@ -647,52 +656,56 @@ func (p *Parser) parseForStatement(scope *scope) Node {
 }
 
 func (p *Parser) parseWhileStatement(scope *scope) Node {
-	tok := p.cur
+	result := &While{}
+	result.Token = p.cur
 	p.advance() // advance past WHILE token
-	scope = newInnerScope(scope)
-	condition := p.parseCondition(scope)
+	scope = scope.innerScope(result)
+	result.Condition = p.parseCondition(scope)
 	p.advancePastNL()
-	block := p.parseBlock(scope)
+	result.Block = p.parseBlock(scope)
 	p.assertEnd()
 	p.advancePastNL()
-	return &While{
-		ConditionalBlock{
-			Token:     tok,
-			Condition: condition,
-			Block:     block,
-		},
+	return result
+}
+
+func inLoop(s *scope) bool {
+	for ; s != nil; s = s.outer {
+		if _, ok := s.block.(*While); ok {
+			return true
+		}
 	}
+	return false
 }
 
 func (p *Parser) parseIfStatement(scope *scope) Node {
-	ifStmt := &If{Token: p.cur}
-	ifStmt.IfBlock = p.parseIfConditionalBlock(scope)
+	result := &If{Token: p.cur}
+	result.IfBlock = p.parseIfConditionalBlock(result, scope)
 	// else if blocks
 	for p.cur.TokenType() == lexer.ELSE && p.peek.TokenType() == lexer.IF {
 		p.advance() // advance past ELSE token
-		elseIfBlock := p.parseIfConditionalBlock(scope)
-		ifStmt.ElseIfBlocks = append(ifStmt.ElseIfBlocks, elseIfBlock)
+		elseIfBlock := p.parseIfConditionalBlock(result, scope)
+		result.ElseIfBlocks = append(result.ElseIfBlocks, elseIfBlock)
 	}
 	// else block
 	if p.cur.TokenType() == lexer.ELSE {
 		p.advance() // advance past ELSE token
 		p.assertEOL()
 		p.advancePastNL()
-		ifStmt.Else = p.parseBlock(newInnerScope(scope))
+		result.Else = p.parseBlock(scope.innerScope(result))
 	}
 	p.assertEnd()
 	p.advancePastNL()
-	return ifStmt
+	return result
 }
 
-func (p *Parser) parseIfConditionalBlock(scope *scope) *ConditionalBlock {
-	tok := p.cur
+func (p *Parser) parseIfConditionalBlock(ifStmt *If, scope *scope) *ConditionalBlock {
+	result := &ConditionalBlock{Token: p.cur}
 	p.advance() // advance past IF token
-	scope = newInnerScope(scope)
-	condition := p.parseCondition(scope)
+	scope = scope.innerScope(ifStmt)
```

Comment:
not sure this is right. The block should be the conditional block, not the if block. the if has multiple blocks and they do not share the scope. they each have their own.

Reply:
a new scope is created, each time, just always containing the same outer block == ifStmt.
It is confusing I can see this, but otherwise it doesn't line up with the else statement.

I cannot have the block create its own scope, because in the case of the While `.block` has to be the While statement (I suppose I could, but again unnecessary extra scoping).

I've updated this now so that the conditional block creates a new block (which means there is an unnecessary extra scope for the while, but I guess that's ok).

Reply:
I still feel a bit wrong about this because of the asymmetry between the else branch and the if | if-else branches.